### PR TITLE
fix: handle non-JSON API responses in AD sync action

### PR DESCRIPTION
**Pull request by tfanouillere-sekoia**
## Description

Fixes https://github.com/SekoiaLab/platform/issues/84255

## Summary by Sourcery

Handle and surface non-JSON API responses in the AD asset synchronization action to avoid crashes and return consistent results.

Bug Fixes:
- Avoid raising JSON decode errors by gracefully handling non-JSON responses from asset management APIs and logging detailed error context.
- Ensure the AD sync action returns an empty data set and stops processing when API calls fail or return unexpected formats instead of proceeding with invalid data.

Enhancements:
- Refactor JSON parsing into a shared helper for GET and POST asset management requests to centralize error handling.

Tests:
- Add a test verifying that a non-JSON response from the assets endpoint sets an action error and returns an empty data payload.

### CODE REVIEW

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 1]
```diff
@@ -40,6 +40,45 @@ def session(self) -> requests.Session:
...
+        try:
+            return response.json()
+        except ValueError:
+            response_text = response.text.strip() or "<empty response>"
+            content_type = response.headers.get("Content-Type", "unknown")
+            self.error(
+                f"Expected JSON response for {request_name}: {response.url} "
+                f"with status code {response.status_code}, content type {content_type}, body: {response_text}"
+            )
+            return None
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py:82]: :+1:

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 2]
```diff
@@ -40,6 +40,45 @@ def session(self) -> requests.Session:
...
+        """Wrap merge_assets with error checking. Returns True if successful."""
+        previous_error_message = self.error_message
+        self.merge_assets(destination=destination, sources=sources)
+        return self.error_message == previous_error_message
+
+    def _safe_put_request(self, endpoint: str, json_data: str) -> bool:
+        """Wrap put_request with error checking. Returns True if successful."""
+        previous_error_message = self.error_message
+        self.put_request(endpoint=endpoint, json_data=json_data)
+        return self.error_message == previous_error_message
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Hello @tfanouillere-sekoia,
What is the purpose of these methods?
Thank you

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 3]
```diff
@@ -94,54 +126,56 @@ def merge_assets(self, destination: str, sources: List[str]) -> None:
...
 
-        # Normalize user_ad_data to a list for uniform processing
         if not isinstance(user_ad_data, list):
             user_ad_data = [user_ad_data]
 
-        # Extract and validate configuration
         if not self.base_url or not self.api_key:
-            self.error("Configuration must include 'base_url' and 'api_key'.")
+            self.error("Configuration must include base_url and api_key.")
+            return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Returning a result when facing an error is not recommended. It may lead the playbook to discard the error.
```suggestion
            return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 4]
```diff
@@ -94,54 +126,56 @@ def merge_assets(self, destination: str, sources: List[str]) -> None:
...
 
         asset_name_field = asset_conf.get("asset_name_field")
         if not asset_name_field:
-            self.error("Configuration must include 'asset_name_field'.")
+            self.error("Configuration must include asset_name_field.")
+            return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here
```suggestion
            return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 5]
```diff
@@ -94,54 +126,56 @@ def merge_assets(self, destination: str, sources: List[str]) -> None:
...
-        responses = []  # To collect responses for each user_ad_data item
-
         for index, single_user_ad_data in enumerate(user_ad_data, start=1):
             self.log(f"Processing user_ad_data item {index}/{len(user_ad_data)}")
 
             asset_name = single_user_ad_data.get(asset_name_field)
             if not asset_name:
-                self.error(f"User AD data does not contain the asset_name_field: '{asset_name_field}'.")
+                self.error(f"User AD data does not contain the asset_name_field: {asset_name_field}.")
+                return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here
```suggestion
            return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 6]
```diff
@@ -94,54 +126,56 @@ def merge_assets(self, destination: str, sources: List[str]) -> None:
...
 
             assert isinstance(asset_name, str)
 
-            # Search for asset by name
-            asset_name_json = self.get_assets(search_query=asset_name)
+            asset_name_json = self._safe_get_assets(search_query=asset_name)
+            if asset_name_json is None:
+                return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here
```suggestion
            return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 7]
```diff
@@ -94,54 +126,56 @@ def merge_assets(self, destination: str, sources: List[str]) -> None:
...
             found_assets = set()
 
             for prop, keys in detection_properties_config.items():
                 for key in keys:
                     value = single_user_ad_data.get(key)
                     if value:
-                        assets = self.get_assets(search_query=value, also_search_in_detection_properties=True)
+                        assets = self._safe_get_assets(search_query=value, also_search_in_detection_properties=True)
+                        if assets is None:
+                            return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here

```suggestion
                            return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 8]
```diff
@@ -178,41 +212,41 @@ def run(self, arguments: dict) -> Dict[str, List[Dict[str, Any]]]:
...
-                    # Ensure asset_uuid is in found_assets
                     found_assets.add(asset_uuid)
 
-                    # Remove destination_asset from sources to merge
                     sources_to_merge = list(found_assets - {destination_asset})
-
                     if sources_to_merge:
-                        self.merge_assets(destination=destination_asset, sources=sources_to_merge)
+                        if not self._safe_merge_assets(destination=destination_asset, sources=sources_to_merge):
+                            return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here

```suggestion
                            return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 9]
```diff
@@ -178,41 +212,41 @@ def run(self, arguments: dict) -> Dict[str, List[Dict[str, Any]]]:
...
 
                     endpoint = f"v2/asset-management/assets/{destination_asset}"
                     self.log(f"PUT request: {endpoint} and payload asset is {json_payload_asset}")
-                    self.put_request(endpoint=endpoint, json_data=json_payload_asset)
+                    if not self._safe_put_request(endpoint=endpoint, json_data=json_payload_asset):
+                        return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here

```suggestion
                        return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 10]
```diff
@@ -178,41 +212,41 @@ def run(self, arguments: dict) -> Dict[str, List[Dict[str, Any]]]:
...
                 else:
                     self.error(f"Unexpected asset name search response: {asset_name_json}")
+                    return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here
```suggestion
                    return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 11]
```diff
@@ -178,41 +212,41 @@ def run(self, arguments: dict) -> Dict[str, List[Dict[str, Any]]]:
...
-                # Asset does not exist, create it
                 created_asset = True
-
-                # Create the asset
                 payload_asset["community_uuid"] = community_uuid
                 create_response = self.post_request(
                     endpoint="v2/asset-management/assets", json_data=json.dumps(payload_asset)
                 )
+                if create_response is None:
+                    return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here

```suggestion
                    return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 12]
```diff
@@ -178,41 +212,41 @@ def run(self, arguments: dict) -> Dict[str, List[Dict[str, Any]]]:
...
+
                 destination_asset = create_response.get("uuid", "")
                 if destination_asset == "":
-                    self.error("Asset creation response does not contain 'uuid'.")
+                    self.error("Asset creation response does not contain uuid.")
+                    return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: same here

```suggestion
                    return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 13]
```diff
@@ -178,41 +212,41 @@ def run(self, arguments: dict) -> Dict[str, List[Dict[str, Any]]]:
...
 
-                # Merge found assets into the new asset
                 sources_to_merge = list(found_assets)
                 if sources_to_merge:
-                    self.merge_assets(destination=destination_asset, sources=sources_to_merge)
+                    if not self._safe_merge_assets(destination=destination_asset, sources=sources_to_merge):
+                        return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]: Same here

```suggestion
                        return
```

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 14]
```diff
@@ -178,41 +212,41 @@ def run(self, arguments: dict) -> Dict[str, List[Dict[str, Any]]]:
...
             else:
                 self.error(f"Unexpected asset name search response: {asset_name_json}")
+                return self._abort(responses)
```
Comment (squioc) [inline Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py]:
```suggestion
                return
```

**Review by squioc — COMMENTED** [review]
Hello, Thank you for the work.
Some changes to apply.
Can you also update the version number in `manifest.json` and add an entry in the `CHANGELOG.md`? Thank you

File: Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py [hunk 15]
```diff
@@ -40,6 +56,30 @@ def session(self) -> requests.Session:
...
+                str(error_message),
+                level="error",
+                request_name=request_name,
+                url=response.url,
+                status_code=response.status_code,
+                content_type=content_type,
+                body=response_text,
+            )
+            raise error_message from exc
+
```
[the remark this answers is not in the document]
Comment (mchupeau-sk) [reply Sekoia.io/sekoiaio/operation_center/synchronize_assets_with_ad.py:83]: Hello @tfanouillere-sekoia, this comment of copilot seems to be interesting, can you go deeper on this proposition :grin:

**Review by mchupeau-sk — APPROVED** [review]
LGFM
